doc/devdocs/locks.rst

```diff
@@ -0,0 +1,110 @@
+.. currentmodule:: Base
+
+****************************************************
+Proper maintenance and care of multi-threading locks
+****************************************************
+
+The following strategies are used to ensure that the code is dead-lock free:
+
+ 1. structure code such that only one lock will need to be acquired at a time
+
+ 2. always acquire shared locks in the same order, as given by the table below
+
+ 3. avoid constructs that expect to need unrestricted recursion
+
+Below are all of the locks that exist in the system and the mechanism in use
+that prevents them from being a potential deadlock.
+
+The following are definitely leaf locks (level 1), and must not try to acquire any other lock:
+
+   * safepoint
+
+       note that this lock is acquired implicitly by ``JL_LOCK`` and ``JL_UNLOCK``.
+       use the ``_NOGC`` variants to avoid that for level 1 locks.
```

Comment:
I think I had this in the comment, but maybe it's also worth mentioning here that a `_NOGC` lock is not allowed to do any allocation or hit any safepoints. Or in the word you use here, any safepoint including allocation, enable/disable GC, exception frame enter and restore, implicitly acquires the safepoint lock.